### PR TITLE
Download input and upload output directly, as opposed to using FUSE

### LINKED ISSUE #701: Help to locate the performance bottleneck of Pachyderm
Refer to [https://github.com/pachyderm/pachyderm/issues/699](url) . It seems that I didn't give a clear expression about performance problem, as well as the speculation about FUSE is not correct. So I do some comparison test today and repost this issue.

I run the word count sample in which data is come from the wiki page just like the official sample mentioned. 

```
"cmd": [ "wget",
        "-e", "robots=off",
        "--recursive",
        "--level", "1",
        "--adjust-extension",
        "--no-check-certificate",
        "--no-directories",
        "--directory-prefix",
        "/pfs/out",
        "https://en.wikipedia.org/wiki/Main_Page"
```

First Equipment: I use AWS m4.Xlarge vm + ubuntu14.04 + Kubernetes 1.2.2 + Pachyderm 1.1.0 to run the wourd count example. Note that there is only one node. The whole process uses 29 minutes and the reduce process uses 24 minutes.

```
ID                                 OUTPUT                                              STARTED             DURATION             STATE
b80331f28ac956ba308011b7ef969054   wordcount_reduce/605aa77a63f4456f867f2c476aced43d   59 minutes ago      24 minutes           success
105a07fead7878072e178a9440e5134f   wordcount_map/366ed6c782024fe0a58e5b43346ea70f      About an hour ago   5 minutes            success
cdeca10ac3f24292824a5c878542517b   wordcount_input/30e1f6d2f41a42eca350f76f15722d8b    About an hour ago   18 seconds           success
```

Second Equipment: I use AWS m4.XLarge + ubuntu14.04 to run the example directly and locally, of which the CPU number and memory size is the same with that of Pachyderm node. The  map.go  program and reduce shell script used in Pachyderm sample are also involved in the Local server processing. Finally, the local server uses about 2.5 minutes to finish the task. The reduce task takes the major part. See my time statistical result here.

```
begin time: Fri Aug  5 07:04:26 2016
cat: ./map: Is a directory
end time: Fri Aug  5 07:06:49 2016
```

Third Equipment: I use AWS m4.XLarge + ubuntu14.04 + Hadoop2.7.2 to run the example. A typical wordcount java program in Hadoop is used to be compared with Pachyderm Pipeline. Unfortunately, Hadoop uses less than 10 seconds although I set the Hadoop as the single-node-non-distribution mode, i.e only one Java process is used.

In conclusion, since the Hadoop Word Count use a different method to realize counting, also , Hadoop is a mature data platform. Please just forget it. Let's focus on the performance difference between local server and Pachyderm Pipeline (2.5minutes vs 30 minutes) to find the performance bottleneck.
Could you give me some idea about the performance problem? I have excluded some impossible reasons.
1. Docker Pull : The docker image is build before experiment. And the docker load is so quick.
2. Sources shortage: Don't be worried about that. The 300 GB disk only spends less than 3%. All of the `tmpfs` has a large space. The memory size is also enough for that.
3. Pipeline Programming : All of the execute command follow the example guide strictly.
4. Service Down : I set a daemon to check whether there is any kubernetes or pachyderm service down during the process. And all services work normally.

Several speculations are given:

1 Filesystem Problem: For a simple computational task, Pachyderm doesn't meet a fatal performance problem.  But in word count example, there are many documents (exactly one word one file), does the frequent file write weaken the performance? Or the docker operation on file is slower than local host?

2  Message Send:  The result of `wordcount_map` is needed to be sent to `wordcount_reduce`. As I observe, this process is not done by memory, I am not sure it is fulfilled by DISK file or RESTful API.  Probably this step would cause a time loss.

### PR DESCRIPTION
**Pull request by derekchiang**
This PR makes it so that PPS no longer uses FUSE for job execution.  Rather, it downloads input data directly to disk, and writes output data directly to disk before uploading it.  In [a benchmark](https://github.com/pachyderm/pachyderm/blob/direct-io/src/server/pachyderm_bench_test.go#L105), this shortens job execution times by 4x.

Fixes #701, #777, #755 #780, #781

Caveats:

1. Currently we don't handle the failure mode where a node doesn't have enough disk space to store input/output data.  This will be handled in a future PR.

### CODE REVIEW

File: src/server/cmd/job-shim/main.go [hunk 1]
```diff
@@ -54,6 +85,27 @@ func do(appEnvObj interface{}) error {
 				return err
 			}
 
+			// Start sending ContinuePod to PPS to signal that we are alive
+			go func() {
+				tick := time.Tick(10 * time.Second)
+				for {
+					<-tick
```
Comment (jdoliner) [inline src/server/cmd/job-shim/main.go]: Superficial but I would write the above 2 lines as:

```go
for <-tick {
```

File: src/server/cmd/job-shim/main.go [hunk 2]
```diff
@@ -54,6 +85,27 @@ func do(appEnvObj interface{}) error {
...
+						&ppsserver.ContinuePodRequest{
+							ChunkID: response.ChunkID,
+							PodName: appEnv.PodName,
+						},
+					)
+					if err != nil {
+						lion.Errorf("error from ContinuePod: %s", err.Error())
+					}
+					if res != nil && res.Exit {
+						os.Exit(0)
```
Comment (jdoliner) [inline src/server/cmd/job-shim/main.go]: This seems like a potential bug, `os.Exit` will cause the program to exit immediately bypassing all deferred function calls. There's 1 later in this function that unmounts the mountpoint which won't run if we hit this line. I think that could potentially cause the pod to not clean up correctly, I've seen leaked fuse mounts cause some problems like that.

File: src/server/cmd/job-shim/main.go (hunk 1 above)
Comment (derekchiang) [inline src/server/cmd/job-shim/main.go]: Good point.

File: src/server/cmd/job-shim/main.go (hunk 2 above)
Comment (derekchiang) [inline src/server/cmd/job-shim/main.go]: Good point!

**Comment by jdoliner** [comment]
How hard would it be to get rid of the `fuse` in `/pfs/fuse/prev`? That seems like a kind of ugly wart to me and like it would only require a small modification to the fuse layer to allow mounting a single commit of a single repo as a single directory rather than having an extra layer of directories in between

File: src/server/cmd/job-shim/main.go (hunk 1 above)
Comment (derekchiang) [inline src/server/cmd/job-shim/main.go]: Actually, that doesn't work cuz `<-tick` is not boolean.

**Comment by derekchiang** [comment]
@jdoliner comments addressed

**Comment by derekchiang** [comment]
Also fixes #573 I think.

**Comment by jdoliner** [comment]
1epkgz

**Pull request merged**